src/mlpack/core/optimizers/cne/cne_impl.hpp

```diff
@@ -0,0 +1,202 @@
+/**
+ * @file cne_impl.hpp
+ * @author Marcus Edel
+ * @author Kartik Nighania
+ *
+ * Conventional Neural Evolution
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_CNE_CNE_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_CNE_CNE_IMPL_HPP
+
+#include "cne.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+CNE::CNE(
+    const size_t populationSize,
+    const size_t maxGenerations,
+    const double mutationProb,
+    const double mutationSize,
+    const double selectPercent,
+    const double finalValue,
+    const double fitnessHist) :
+    populationSize(populationSize),
+    maxGenerations(maxGenerations),
+    mutationProb(mutationProb),
+    mutationSize(mutationSize),
+    selectPercent(selectPercent),
+    finalValue(finalValue),
+    fitnessHist(fitnessHist)
+{ /* Nothing to do here. */ }
+
+//! Optimize the function
+template<typename DecomposableFunctionType>
+double CNE::Optimize(
+    DecomposableFunctionType& function,
+    arma::mat& iterate)
+{
+  // Get the number of functions to iterate
+  const size_t numFun = function.NumFunctions();
+
+  // Set the population size and fill random values [0,1]
+  population = arma::randu(populationSize, iterate.n_rows);
+
+  // initialize helper variables
+  fitnessValues.set_size(populationSize);
+  double fitness = 0;
+  double lastBestFitness;
+
+  Log::Info << "CNE initialized successfully. Optimization started"
+      << std::endl;
+
+  // Iterate till max number of generations
+  for (size_t gen = 1; gen <= maxGenerations; gen++)
+  {
+    // calculate fitness values of all candidates
+    for (size_t i = 0; i < populationSize; i++)
+    {
+       // select a candidate and insert the parameters in the function
+       iterate = population.row(i).t();
+
+       // find the fitness
+       for (size_t j = 0; j < numFun; j++)
+         fitness += function.Evaluate(iterate, j);
+
+       // Save fitness values
+       fitnessValues[i] = fitness;
+       fitness = 0;
+    }
+
+      Log::Info << "Generation number: " << gen << " best fitness = "
+          << fitnessValues.min() << std::endl;
+
+      // create the next generation of species
+      Reproduce();
+
+      // check for termination
+      if (finalValue != DBL_MIN && finalValue >= fitnessValues.min())
+      {
+          Log::Info << "Terminating. Given fitness criteria " << finalValue
+              << " > " << fitnessValues.min() << std::endl;
+          break;
+      }
+
+      // check for termination
+      if (fitnessHist != DBL_MIN && gen != 1 &&
```

Comment:
Instead of checking against ``gen == 1`` we could lastBestFitness fitness at the beginning by iterating through the values with the initial parameter.